addon/authenticators/oauth2-implicit-grant.js

```diff
@@ -0,0 +1,118 @@
+   @return {Ember.RSVP.Promise} A promise that when it resolves results in the session becoming or remaining authenticated
+   @public
+   */
+  restore(data) {
+    return new RSVP.Promise((resolve, reject) => {
+      if (!this._validateToken(data)) {
+        return reject();
+      }
+
+      return resolve(data);
+    });
+  },
+
+  /**
+   Authenticates the session using the specified `response`
+   (see https://tools.ietf.org/html/rfc6749?hmm=lol#section-4.2.2); if no response
+   ist set the response will be parsed from `window.location.hash`.
+
+   __If the access token is valid and thus authentication succeeds, a promise that
+   resolves with the access token is returned__, otherwise a promise that rejects
+   with the error as returned by the server is returned
+   (see https://tools.ietf.org/html/rfc6749?hmm=lol#section-4.2.2.1).
+
+   @method authenticate
+   @param {Object} response The response
+   @return {Ember.RSVP.Promise} A promise that when it resolves results in the session becoming authenticated
+   @public
+   */
+  authenticate(response) {
+    return new RSVP.Promise((resolve, reject) => {
+      if (isEmpty(response)) {
+        response = this._parseResponse(window.location.hash);
+      }
+
+      if (response.error) {
+        reject(response);
+      } else if (!this._validateToken(response)) {
+        reject({
+          error: 'invalid_token',
+          // jscs:disable requireCamelCaseOrUpperCaseIdentifiers
+          error_description: 'The received token is invalid'
+          // jscs:enable requireCamelCaseOrUpperCaseIdentifiers
+        });
+      } else {
+        resolve(response);
+      }
+    });
+  },
+
+  /**
+   This method simply returns a resolving promise.
+
+   @method invalidate
+   @return {Ember.RSVP.Promise} A promise that when it resolves results in the session being invalidated
+   @public
+   */
+  invalidate() {
+    return RSVP.Promise.resolve();
+  },
+
+  _validateToken(token) {
+    // see https://tools.ietf.org/html/rfc6749#section-4.2.2
+
+    // jscs:disable requireCamelCaseOrUpperCaseIdentifiers
+    return !isEmpty(token) &&
+      !isEmpty(token.access_token) &&
+      !isEmpty(token.state) &&
+      !isEmpty(token.token_type);
+    // jscs:enable requireCamelCaseOrUpperCaseIdentifiers
+  },
+
+  _parseResponse(locationHash) {
+    let params = {};
+    const query = locationHash.substring(locationHash.indexOf('?'));
+    const regex = /([^#?&=]+)=([^&]*)/g;
+    let match;
+
+    // decode all parameter pairs
+    while ((match = regex.exec(query)) !== null) {
+      params[decodeURIComponent(match[1])] = decodeURIComponent(match[2]);
+    }
+
+    // parse eventual error description
+    // jscs:disable requireCamelCaseOrUpperCaseIdentifiers
+    if (params.error_description) {
+      params.error_description = params.error_description.replace(/\+/g, ' ');
+    }
+    // jscs:enable requireCamelCaseOrUpperCaseIdentifiers
+
+    return params;
+  }
```

Comment:
To be secure authentication servers must append the token after the fragment e.g. `/callback#access_token=1234`. In that sense we cannot rely on Embers abstractions here.

See: https://tools.ietf.org/html/rfc6749#section-4.2.2.

I'd rather suggest to keep it in the authenticator so that the route callback mixin is not mandatory for using the authenticator.